# 如何处理参考基因组的非染色体序列？

原文链接：[ (How to) Map reads to a reference with alternate contigs like GRCh38](https://software.broadinstitute.org/gatk/documentation/article?id=8017)

这篇教程大部分内容在其它教程都讲解过，不需要重复翻译。需要搞懂的是为什么要把非染色体序列加入参考基因组，以及后续的变异寻找过程该如何对待非染色体区域。



[![img](https://us.v-cdn.net/5019796/uploads/FileUpload/65/a0f09aad5f351a1322f7c1b19ec5d9.png)](https://us.v-cdn.net/5019796/uploads/FileUpload/65/a0f09aad5f351a1322f7c1b19ec5d9.png)This exploratory tutorial provides instructions and example data to map short reads to a reference genome with alternate haplotypes. Instructions are suitable for indexing and mapping reads to GRCh38.

► If you are unfamiliar with terms that describe reference genome components, or GRCh38 alternate haplotypes, take a few minutes to study the *Dictionary* entry [Reference Genome Components](https://software.broadinstitute.org/gatk/documentation/article?id=7857).

► For an introduction to GRCh38, see [Blog#8180](https://software.broadinstitute.org/gatk/blog?id=8180).

Specifically, the tutorial uses BWA-MEM to index and map simulated reads for three samples to a mini-reference composed of a GRCh38 chromosome and alternate contig (**sections 1–3**). We align in an alternate contig aware (alt-aware) manner, which we also call alt-handling. This is the main focus of the tutorial.



## 5. How can I tell whether I should consider an alternate haplotype?

[![img](https://us.v-cdn.net/5019796/uploads/FileUpload/30/3d6bbf9a4c67674e1ebea0308bdd3f.png)](https://us.v-cdn.net/5019796/uploads/FileUpload/30/3d6bbf9a4c67674e1ebea0308bdd3f.png)We consider this question only for our GPI locus, a locus we know has an alternate contig in the reference. Here we use the term *locus* in its biological sense to refer to a contiguous genomic region of interest. The three samples give the alignment and coverage profiles shown on the right.

What is immediately apparent from the IGV screenshot is that the scenarios that include the alternate haplotype give a distinct pattern of variant sites to the primary assembly much like a fingerprint. These variants are predominantly heterozygous or homozygous. Looking closely at the 3' region of the locus, we see some alignment coverage anomalies that also show a distinct pattern. The coverage in some of the highly diverged region in the primary assembly drops while in others it increases. If we look at the origin of simulated reads in one of the excess coverage regions, we see that they are from two different regions of the alternate contig that suggests duplicated sequence segments within the alternate locus.

The variation pattern and coverage anomalies on the primary locus suggest an alternate haplotype may be present for the locus. We can then confirm the presence of aligned reads, both supplementary and primary, on the alternate locus. Furthermore, if we count the alignment records for each region, e.g. using `samtools idxstats`, we see the following metrics.

```
                        ALT/ALT     PA/ALT     PA/PA   
chr19                     10005      10006     10000     
chr19_KI270866v1_alt       1407        799         0      
```

The number of alignments on the alternate locus increases proportionately with alternate contig dosage. All of these factors together suggest that the sample presents an alternate haplotype.

### 5.1 Discussion of variant calls for tutorial_8017

The three-sample variant callset gives 54 sites on the primary locus and two additional on the alternate locus for 56 variant sites. All of the eight SNP alleles we introduced are called, with six called on the primary assembly and two called on the alternate contig. Of the 15 expected genotype calls, four are incorrect. Namely, four PAALT calls that ought to be heterozygous are called homozygous variant. These are two each on the primary assembly and on the alternate contig in the region that is highly divergent.

► Our production pipelines use genomic intervals lists that exclude GRCh38 alternate contigs from *variant calling*. That is, variant calling is performed only for contigs of the primary assembly. This calling on even just the primary assembly of GRCh38 brings improvements to analysis results over previous assemblies. For example, if we align and call variants for our simulated reads on GRCh37, we call 50 variant sites with identical QUAL scores to the equivalent calls in our GRCh38 callset. However, this GRCh37 callset is missing six variant calls compared to the GRCh38 callset for the 42 kb locus: the two variant sites on the alternate contig and *four variant sites on the primary assembly*.

Consider the example variants on the primary locus. The variant calls from the primary assembly include 32 variant sites that are strictly homozygous variant in ALTALT *and* heterozygous variant in PAALT. The callset represents only those reads from the ALT *that can be mapped to the primary assembly*.

In contrast, the two variants in regions whose reads *can only map to the alternate contig* are absent from the primary assembly callset. For this simulated dataset, the primary alignments present on the alternate contig provide enough supporting reads that allow HaplotypeCaller to call the two variants. However, these variant calls have *lower-quality annotation metrics* than for those simulated in an equal manner on the primary assembly. We will get into why this is in **section 6**.

Additionally, for our PAALT sample that is heterozygous for an alternate haplotype, the genotype calls in the highly divergent regions are inaccurate. These are called homozygous variant on the primary assembly and on the alternate contig when in fact they are heterozygous variant. These calls have lower genotype scores `GQ` as well as lower allele depth `AD` and coverage `DP`. The table below shows the variant calls for the introduced SNP sites. In blue are the genotype calls that should be heterozygous variant but are instead called homozygous variant. [![img](https://us.v-cdn.net/5019796/uploads/FileUpload/0a/bc1a0d986fcf6087058c9ce46551bf.png)](https://us.v-cdn.net/5019796/uploads/FileUpload/0a/bc1a0d986fcf6087058c9ce46551bf.png)

Here is a command to select out the intentional variant sites that uses [SelectVariants](https://software.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_variantutils_SelectVariants.php):

```
java -jar GenomeAnalysisTK.jar -T SelectVariants \
    -R chr19_chr19_KI270866v1_alt.fasta \
    -V multisample.vcf -o multisample_selectvariants.vcf \
    -L chr19:34,383,500 -L chr19:34,389,485 -L chr19:34,391,800 -L chr19:34,392,600 \
    -L chr19_KI270866v1_alt:32,700 -L chr19_KI270866v1_alt:38,700 \
    -L chr19_KI270866v1_alt:41,700 -L chr19_KI270866v1_alt:42,700 \
    -L chr19:34,383,486 -L chr19_KI270866v1_alt:32,714 
```

[back to top](https://software.broadinstitute.org/gatk/documentation/topic?name=tutorials#top)

------

## 6. My locus includes an alternate haplotype. How can I call variants on alt contigs?

If you want to call variants on alternate contigs, consider additional data processing that overcome the following problems.

- Loss of alignments from filtering of overclipped reads.
- HaplotypeCaller's filtering of alignments whose mates map to another contig. Alt-handling produces many of these types of reads on the alternate contigs.
- Zero MAPQ scores for alignments that map to two or more alternate contigs. HaplotypeCaller excludes these types of reads from contributing to evidence for variation.

Let us talk about these in more detail.

Ideally, if we are interested in alternate haplotypes, then we would have ensured we were using the most up-to-date analysis reference genome sequence with the latest patch fixes. Also, whatever approach we take to align and preprocess alignments, if we filter any reads as putative contaminants, e.g. with MergeBamAlignment's option to unmap cross-species contamination, then at this point we would want to fish back into the unmapped reads pool and pull out those reads. Specifically, these would have an `SA` tag indicating mapping to the alternate contig of interest and an `FT` tag indicating the reason for unmapping was because MergeBamAlignment's `UNMAP_CONTAMINANT_READS` option identified them as cross-species contamination. Similarly, we want to make sure not to include HaplotypeCaller's `--read_filter OverclippedRead` option that we use in the first workflow.

[![img](https://us.v-cdn.net/5019796/uploads/FileUpload/28/f8c0855ffef62382f0e96e53a82977.png)](https://us.v-cdn.net/5019796/uploads/FileUpload/28/f8c0855ffef62382f0e96e53a82977.png)As **section 5.1** shows, variant calls on the alternate contig are of low quality--they have roughly an order of magnitude lower QUAL scores than what should be equivalent variant calls on the primary assembly.

For this exploratory tutorial, we are interested in calling the introduced SNPs with equivalent annotation metrics. Whether they are called on the primary assembly or the alternate contig and whether they are called homozygous variant or heterozygous--let's say these are less important, especially given pinning certain variants from highly homologous regions to one of the loci is nigh impossible with our short reads. To this end, we will use the second workflow shown in the workflows diagram. However, because this solution is limited, we present a third workflow as well.

► We present these workflows solely for exploratory purposes. They do not represent any production workflows.

**Tutorial_8017_toSE** uses the processed BAM from our first workflow and allows for calling on singular alternate contigs. That is, the workflow is suitable for calling on alternate contigs of loci with only a single alternate contig like our GPI locus. **Tutorial_8017_postalt** uses the aligned SAM from the first workflow before processing, and requires separate processing before calling. This third workflow allows for calling on all alternate contigs, even on HLA loci that have numerous contigs per primary locus. However, the callset will not be parsimonious. That is, each alternate contig will greedily represent alignments and it is possible the same variant is called for all the alternate loci for a given primary locus as well as on the primary locus. It is up to the analyst to figure out what to do with the resulting calls.

[![img](https://us.v-cdn.net/5019796/uploads/FileUpload/fc/2834e6593da374296a205f33d117ac.png)](https://us.v-cdn.net/5019796/uploads/FileUpload/fc/2834e6593da374296a205f33d117ac.png)The reason for the divide in these two workflows is in the way BWA assigns mapping quality scores (MAPQ) to multimapping reads. Postalt-processing becomes necessary for loci with two or more alternate contigs because the shared alignments between the primary locus and alternate loci will have zero MAPQ scores. Postalt-processing gives non-zero MAPQ scores to the alignment records. The table presents the frequencies of GRCh38 non-HLA alternate contigs per primary locus. It appears that ~75% of non-HLA alternate contigs are singular to ~92% of primary loci with non-HLA alternate contigs. In terms of bases on the primary assembly, of the ~75 megabases that have alternate contigs, ~64 megabases (85%) have singular non-HLA alternate contigs and ~11 megabases (15%) have multiple non-HLA alternate contigs per locus. Our tutorial's example locus falls under this majority.

In both alt-aware mapping and postalt-processing, alternate contig alignments have a predominance of mates that map back to the primary assembly. HaplotypeCaller, for good reason, filters reads whose mates map to a different contig. However, we know that GRCh38 *artificially* represents alternate haplotypes as separate contigs and BWA-MEM *intentionally* maps these mates back to the primary locus. For comparable calls on alternate contigs, we need to include these alignments in calling. To this end, we have devised a temporary workaround.

### 6.1 Variant calls for **tutorial_8017_toSE**

Here we are only aiming for *equivalent calls* with similar annotation values for the two variants that are called on the alternate contig. For the solution that we will outline, here are the results.

[![img](https://us.v-cdn.net/5019796/uploads/FileUpload/81/62cd9b4e7c710dc94ea6c4bdb45db9.png)](https://us.v-cdn.net/5019796/uploads/FileUpload/81/62cd9b4e7c710dc94ea6c4bdb45db9.png)

Including the mate-mapped-to-other-contig alignments bolsters the variant call qualities for the two SNPs HaplotypeCaller calls on the alternate locus. We see the `AD` allele depths much improved for ALTALT and PAALT. Corresponding to the increase in reads, the `GQ` genotype quality and the QUAL score (highlighted in red) indicate higher qualities. For example, the QUAL scores increase from 332 and 289 to 2166 and 1764, respectively. We also see that one of the genotype calls changes. For sample ALTALT, we see a previous *no call* is now a homozygous reference call (highlighted in blue). This hom-ref call is further from the truth than not having a call as the ALTALT sample should not have coverage for this region in the primary assembly.

For our example data, **tutorial_8017**'s callset subset for the primary assembly and **tutorial_8017_toSE**'s callset subset for the alternate contigs together appear to make for a better callset.

What solution did we apply? As the workflow's name *toSE* implies, this approach converts paired reads to single end reads. Specifically, this approach takes the processed and coordinate-sorted BAM from the first workflow and removes the 0x1 *paired* flag from the alignments. Removing the 0x1 flag from the reads allows HaplotypeCaller to consider alignments whose mates map to a different contig. We accomplish this using a modified script of that presented in *Biostars* post <https://www.biostars.org/p/106668/>, indexing with Samtools and then calling with HaplotypeCaller as follows. Note this workaround creates an invalid BAM according to ValidateSamFile. Also, another caveat is that because HaplotypeCaller uses softclipped sequences, any overlapping regions of read pairs will count twice towards variation instead of once. Thus, this step may lead to overconfident calls in such regions.

Remove the 0x1 bitwise flag from alignments

```
samtools view -h altalt_snaut.bam | gawk '{printf "%s\t", $1; if(and($2,0x1))
{t=$2-0x1}else{t=$2}; printf "%s\t" , t; for (i=3; i<NF; i++){printf "%s\t", $i} ; 
printf "%s\n",$NF}'| samtools view -Sb - > altalt_se.bam
```

Index the resulting BAM

```
samtools index altalt_se.bam
```

Call variants in `-ERC GVCF` mode with HaplotypeCaller for each sample

```
java -jar GenomeAnalysisTK.jar -T HaplotypeCaller \
    -R chr19_chr19_KI270866v1_alt.fasta \
    -I altalt_se.bam -o altalt_hc.g.vcf \
    -ERC GVCF --emitDroppedReads -bamout altalt_hc.bam
```

Finally, use GenotypeGVCFs as shown in [section 4](https://software.broadinstitute.org/gatk/documentation/topic?name=tutorials#4)'s command [4.7] for a multisample variant callset. Tutorial_8017_toSE calls 68 variant sites--66 on the primary assembly and two on the alternate contig.

### 6.2 Variant calls for **tutorial_8017_postalt**

BWA's postalt-processing requires the query-grouped output of BWA-MEM. Piping an alignment step with postalt-processing is possible. However, to be able to compare variant calls from an identical alignment, we present the postalt-processing as an *add-on* workflow that takes the alignment from the first workflow.

The command uses the `bwa-postalt.js` script, which we run through `k8`, a Javascript execution shell. It then lists the ALT index, the aligned SAM `altalt.sam` and names the resulting file `> altalt_postalt.sam`.

```
k8 bwa-postalt.js \
    chr19_chr19_KI270866v1_alt.fasta.alt \
    altalt.sam > altalt_postalt.sam
```

[![img](https://us.v-cdn.net/5019796/uploads/FileUpload/a5/3522324635aec94071a9ff688a4aa6.png)](https://us.v-cdn.net/5019796/uploads/FileUpload/a5/3522324635aec94071a9ff688a4aa6.png)The resulting postalt-processed SAM, `altalt_postalt.sam`, undergoes the same processing as the first workflow (commands 4.1 through 4.7) except that (i) we omit `--max_alternate_alleles 3`and `--read_filter OverclippedRead` options for the HaplotypeCaller command like we did in **section 6.1** and (ii) we perform the 0x1 flag removal step from **section 6.1**.

The effect of this postalt-processing is immediately apparent in the IGV screenshots. Previously empty regions are now filled with alignments. Look closely in the highly divergent region of the primary locus. Do you notice a change, albeit subtle, before and after postalt-processing for samples ALTALT and PAALT?

These alignments give the calls below for our SNP sites of interest. Here, notice calls are made for more sites--on the equivalent site if present in addition to the design site (highlighted in the first two columns). For the three pairs of sites that can be called on either the primary locus or alternate contig, the variant site QUALs, the INFO field annotation metrics and the sample level annotation values are identical for each pair.

[![img](https://us.v-cdn.net/5019796/uploads/FileUpload/c4/5f07b27798374175ba40f970e77a62.png)](https://us.v-cdn.net/5019796/uploads/FileUpload/c4/5f07b27798374175ba40f970e77a62.png)

Postalt-processing lowers the MAPQ of primary locus alignments in the highly divergent region that map better to the alt locus. You can see this as a subtle change in the IGV screenshot. After postalt-processing we see an increase in white zero MAPQ reads in the highly divergent region of the primary locus for ALTALT and PAALT. For ALTALT, this effectively cleans up the variant calls in this region at chr19:34,391,800 and chr19:34,392,600. Previously for ALTALT, these calls contained some reads: 4 and 25 for the first workflow and 0 and 28 for the second workflow. After postalt-processing, no reads are considered in this region giving us `./.:0,0:0:.:0,0,0`calls for both sites.

What we omit from examination are the effects of postalt-processing on decoy contig alignments. Namely, if an alignment on the primary assembly aligns better on a decoy contig, then postalt-processing discounts the alignment on the primary assembly by assigning it a zero MAPQ score.

To wrap up, here are the number of variant sites called for the three workflows. As you can see, this last workflow calls the most variants at 95 variant sites, with 62 on the primary assembly and 33 on the alternate contig.

```
Workflow                total    on primary assembly    on alternate contig
tutorial_8017           56       54                      2
tutorial_8017_toSE      68       66                      2
tutorial_8017_postalt   95       62                     33
```

[back to top](https://software.broadinstitute.org/gatk/documentation/topic?name=tutorials#top)

------

### 7. Related resources

- For WDL scripts of the workflows represented in this tutorial, see the [GATK WDL scripts repository](https://github.com/broadinstitute/wdl/tree/develop/scripts/tutorials/gatk).
- To revert an aligned BAM to unaligned BAM, see **Section B** of [Tutorial#6484](https://software.broadinstitute.org/gatk/documentation/article?id=6484).
- To simulate reads from a reference contig, see [Tutorial#7859](https://software.broadinstitute.org/gatk/documentation/article?id=7859).
- *Dictionary* entry [Reference Genome Components](https://software.broadinstitute.org/gatk/documentation/article?id=7857) reviews terminology that describe reference genome components.
- The [GATK resource bundle](https://software.broadinstitute.org/gatk/download/bundle) provides an analysis set GRCh38 reference FASTA as well as several other related resource files.
- As of this writing (August 8, 2016), the SAM format ALT index file for GRCh38 is available only in the [x86_64-linux bwakit download](https://sourceforge.net/projects/bio-bwa/files/bwakit/) as stated in this [bwakit README](https://github.com/lh3/bwa/tree/master/bwakit). The `hs38DH.fa.alt` file is in the `resource-GRCh38` folder. Rename this file's basename to match that of the corresponding reference FASTA.
- For more details on MergeBamAlignment features, see [Section 3C](https://software.broadinstitute.org/gatk/documentation/article?id=6483#step3C) of [Tutorial#6483](https://software.broadinstitute.org/gatk/documentation/article?id=6483).
- For details on the PairedEndSingleSampleWorkflow that uses GRCh38, see [here](https://github.com/broadinstitute/wdl/blob/develop/scripts/broad_pipelines/PairedSingleSampleWf_160720.md).
- See [here](https://samtools.github.io/hts-specs) for VCF specifications.

